feature-libs/product-configurator/rulebased/components/attribute/types/radio-button/configurator-attribute-radio-button.component.ts

```diff
@@ -12,10 +15,41 @@ export class ConfiguratorAttributeRadioButtonComponent
   extends ConfiguratorAttributeSingleSelectionBaseComponent
   implements OnInit
 {
+  uiType = Configurator.UiType;
+
   attributeRadioButtonForm = new FormControl('');
 
-  constructor(protected quantityService: ConfiguratorAttributeQuantityService) {
+  constructor(
+    protected quantityService: ConfiguratorAttributeQuantityService,
+    protected innerComponentsContext: InnerComponentsContext,
+    protected configuratorCommonsService: ConfiguratorCommonsService
+  ) {
     super(quantityService);
+    this.attribute = innerComponentsContext.context.attribute;
+    this.ownerKey = innerComponentsContext.context.configuration.owner.key;
+  }
+
+  onSelect(value: string): void {
+    this.loading$.next(true);
+
+    // const event: ConfigFormUpdateEvent = {
+    //   changedAttribute: {
+    //     ...this.attribute,
+    //     selectedSingleValue: value,
+    //   },
+    //   ownerKey: this.ownerKey,
+    //   updateType: Configurator.UpdateType.ATTRIBUTE,
+    // };
+
+    //this.selectionChange.emit(event);
+    this.configuratorCommonsService.updateConfiguration(
+      this.ownerKey,
+      {
+        ...this.attribute,
+        selectedSingleValue: value,
+      },
+      Configurator.UpdateType.ATTRIBUTE
+    );
```

Comment:
Let me know if I understand correctly the communication model here:
- any input data for this component are passed via the `context`
- but output of this component (e.g. side-effect commands dispatched to services) are communicated directly to the specific services.

This doesn't look bad. At the same time I'm wondering about encapsulating input and output in the context itself (e.g. outputs would be methods of the context). For example the context class `ConfiguratorFormContext`would have a method `updateConfiguration()` (which I guess all those inner components need).

I'm talking here, regardless the method of passing the context class. I'm focusing here on the idea of the context that encapsulates both input and output. What do you think about it? Does it simplify the architecture/usage/brings benefits or is just an unnecessary abstraction to you?

Reply:
Yes, your understanding of the communication model is correct.
Not easy to answer your question. 
First for me it felt like we _should_ let the context also handle the output, but TBH I didn't get it to work and just wanted to assess the overall feasibility. 
Now, when I think about it, I am not 100% sure we need that. Our case seems to suggest it, especially because the update was handled by the form before, but calling the facade service from the sub components even reduces the complexity. And when I think of other use cases, I assume that parent/child components are normally more loosely coupled than in our, quite specific case, meaning less need for such a callback option.

So I tend to think that it's not needed, but might fundamentally miss other options to use that concept.